openerp/addons/base/module/module.py

```diff
@@ -414,31 +414,48 @@ def state_update(self, newstate, states_to_update, level=100):
-        #  - all dependencies satisfied (installed or to be installed),
-        #  - at least one dependency being 'to install'
-        satisfied_states = frozenset(('installed', 'to install', 'to upgrade'))
-        def all_depencies_satisfied(m):
-            states = set(d.state for d in m.dependencies_id)
-            return states.issubset(satisfied_states) and ('to install' in states)
-        to_install_modules = filter(all_depencies_satisfied, uninstalled_modules)
-        to_install_ids = map(lambda m: m.id, to_install_modules)
-
-        # Mark them to be installed.
-        if to_install_ids:
-            self.button_install(cr, uid, to_install_ids, context=context)
+    @api.multi
+    def button_install(self):
+        # domain to select auto-installable (but not yet installed) modules
+        auto_domain = [('state', '=', 'uninstalled'), ('auto_install', '=', True)]
+
+        # determine whether an auto-install module must be installed:
+        #  - all its dependencies are installed or to be installed,
+        #  - at least one dependency is 'to install'
+        install_states = frozenset(('installed', 'to install', 'to upgrade'))
+        def must_install(module):
+            states = set(dep.state for dep in module.dependencies_id)
+            return states <= install_states and 'to install' in states
+
+        modules = self
+        while modules:
+            # Mark the given modules and their dependencies to be installed.
+            modules.state_update('to install', ['uninstalled'])
+
+            # Determine which auto-installable modules must be installed.
+            modules = self.search(auto_domain).filtered(must_install)
+
+        # retrieve the installed (or to be installed) theme modules
+        theme_category = self.env.ref('base.module_category_theme')
+        theme_modules = self.search([
+            ('state', 'in', list(install_states)),
+            ('category_id', 'child_of', [theme_category.id]),
+        ])
+
+        # determine all theme modules that module depends on, including itself
+        def theme_deps(module):
+            deps = module.mapped('dependencies_id.depend_id')
+            return (module & theme_modules) | (deps and deps.mapped(theme_deps))
+
+        # check: the installation is valid if all installed theme modules
+        # correspond to one theme module and all its theme dependencies
+        if 'to install' in theme_modules.mapped('state') and \
+                all(theme_deps(module) != theme_modules for module in theme_modules):
+            raise UserError(_(
+                "You are trying to install non-compatible theme modules:\n%s\n\n" \
+                "You should first uninstall existing theme modules before installing another theme module.\n"
+                "Beware that switching themes may have some impact on your website pages!"
+            ) % (", ".join("%s (%s)" % (module.shortdesc, module.name) for module in theme_modules),))
```

Comment:
Perhaps we can include the themes as a bullet list, and indicate the state of the theme instead of its technical name, e.g. something like this pseudo-code:
```python
state_labels = dict(self.fields_get(['state'])['state']['selection'])
themes_list = ["\n- %s (%s)" % (m.shortdesc, state_labels[m.state]) for m in theme_modules]
raise UserError(_(
    "You are trying to install incompatible theme modules:\n%s\n\n" \
    "Please uninstall your current theme before installing another one.\n"
    "Warning: switching themes may significantly alter the look of your current website pages!"
) % (''.join(themes_list),))
```

We should probably also modify the warning generated by the uninstall wizard to add this warning about the current website if the user is uninstalling a theme.

Thanks!